rs/nns/governance/src/voting.rs

```diff
@@ -22,37 +78,146 @@ impl Governance {
+
+        // We process until voting is finished, and then do any other work that fits into the soft
+        // limit of the current message.  Votes are guaranteed to be recorded before the function
+        // returns, but recent_ballots for neurons might be recorded later in a timer job.  This
+        // ensures we return to the caller in a reasonable amount of time.
+        let mut is_voting_finished = false;
+
+        while !is_voting_finished {
+            // Now we process until we are done or we are over a limit and need to
+            // make a self-call.
+            with_voting_state_machines_mut(|voting_state_machines| {
+                voting_state_machines.with_machine(proposal_id, topic, |machine| {
+                    self.process_machine_until_soft_limit(machine, over_soft_message_limit);
+                    is_voting_finished = machine.is_voting_finished();
+                });
+            });
+
+            // This returns an error if we hit the hard limit, which should basically never happen
+            // in production, but we need a way out of this loop in the worst case to prevent
+            // the canister from being unable to upgrade.
+            if let Err(e) = noop_self_call_if_over_instructions(
+                SOFT_VOTING_INSTRUCTIONS_LIMIT,
+                Some(HARD_VOTING_INSTRUCTIONS_LIMIT),
+            )
+            .await
+            {
+                println!(
+                    "Error in cast_vote_and_cascade_follow, \
+                        voting will be processed in timers: {}",
+                    e
+                );
+                break;
             }
-        };
+        }
+        // We use the time from the beginning of the function to retain the behaviors needed
+        // for wait for quiet even when votes can be processed asynchronously.
+        self.recompute_proposal_tally(proposal_id, voting_started);
+    }
 
-        // Use of thread local storage to store the state machines prevents
-        // more than one state machine per proposal, which limits the overall
-        // memory usage for voting, which will be relevant when this can be used
-        // across multiple messages, which would cause the memory usage to accumulate.
-        VOTING_STATE_MACHINES.with(|vsm| {
-            let mut voting_state_machines = vsm.borrow_mut();
-            let proposal_voting_machine =
-                voting_state_machines.get_or_create_machine(proposal_id, topic);
+    /// Record a neuron vote into the voting state machine, then do nothing else.
+    fn record_neuron_vote(
+        &mut self,
+        proposal_id: ProposalId,
+        voting_neuron_id: NeuronId,
+        vote: Vote,
+        topic: Topic,
+    ) {
+        with_voting_state_machines_mut(|voting_state_machines| {
+            let ballots = &mut self
+                .heap_data
+                .proposals
+                .get_mut(&proposal_id.id)
+                .unwrap()
```

Comment:
These unwraps are probably always safe (in practice), but b/c of interleaving messages, it should be safely accessed, and we need a way to back out without panicking.